### PR TITLE
[css-text] Update line-break tests to match i18n test suite

### PR DESCRIPTION
**Pull request by xfq**
This commit tests ja and zh line-break behavior with testharness.js, replacing PR #9994.

Related issue: https://github.com/w3c/csswg-drafts/issues/2450

### CODE REVIEW

**Review by gsnedders — COMMENTED** [review]
These apply to all new files; haven't checked the correctness of the tests though.

File: css/css-text/i18n/zh/css-text-line-break-zh-cpm-loose.html
Comment (r12a) [reply css/css-text/i18n/zh/css-text-line-break-zh-cpm-loose.html]: @xfq I agree with @gsnedders, and apologise that I should have mentioned that earlier. However, please make these changes only in the files submitted to WPT, because top-level absolute paths don't work well when the tests are used in the i18n test suite.

File: css/css-text/i18n/zh/css-text-line-break-zh-cpm-loose.html
Comment (r12a) [reply css/css-text/i18n/zh/css-text-line-break-zh-cpm-loose.html]: Same here.

**Comment by xfq** [comment]
I'm not a big fan of using the word "[path](https://www.gnu.org/prep/standards/html_node/GNU-Manuals.html#GNU-Manuals)", but anyway...

I've updated the PR.

File: css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html
Comment (mrego) [inline css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html]: Nit: Please use 2 lines for this, it's hard to read.

File: css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html
Comment (mrego) [inline css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html]: Nit: Like @gsnedders I don't see the need to split this in 2 loops, but I don't have a strong opinion about it if you want to keep like it.

File: css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html
Comment (mrego) [inline css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html]: Nit: I guess you can directly use `document.body`

**Review by mrego — COMMENTED** [review]
@kojiishi or @frivoal should be the ones reviewing this PR,
as I don't know the topic well enough.

Anyway I left a bunch of comments on one of the new files (mostly nits that you can feel free to ignore).

It seems that the JavaScript code is repeated in all the files (with some minor differences),
wouldn't be better to have a helper method defined in a separated file that would be reused in all the tests?

File: css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html
Comment (r12a) [reply css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html]: I find it easier to read and and easier to spot errors when i indent my code as shown here (in Python style).

File: css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html
Comment (r12a) [reply css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html]: Again, for me it improves readability and speeds up editing to omit semicolons.

Comment (mrego) [reply css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html]: Most of the code I've seen in WPT uses semicolons, anyway if you don't want to use them you should be consistent (check the 2nd loop where you're using semicolons while you're not using them in the rest of the file).

File: css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html
Comment (r12a) [reply css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html]: Indeed, understood. But for me those unindented }s make it harder to see the clean lines of each set of related items. I dislike the common approach so much that i explicity change them all, since i use DreamWeaver and it automatically does what i don't want (and provides no way to adapt its behaviour that i can see).

File: css/css-text/i18n/ja/css-text-line-break-ja-in-loose.html
Comment (kojiishi) [inline css/css-text/i18n/ja/css-text-line-break-ja-in-loose.html]: U+10AF6 is probably not in 'mplus-1p-regular' and that not testable. We could add a font that has a glyph, or use test font such as CSS Orientation test font, but this isn't a CJK character, probably not worth to test?

Appear in all "*-in-*" files.

**Review by kojiishi — COMMENTED** [review]
lgtm w/one comment above. Also please make sure @mrego happy before merging.

File: css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html
Comment (kojiishi) [reply css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html]: Yeah, this is the one I asked to do so in pre-review with Richard. Reading |offsetLeft| forces a layout, so repeating "change DOM" + "reading |offsetLeft|" requires multiple layouts. The character list in each test ended up in not so large number of lines, so 1 loop might be ok for these tests, but generally data-driven tests should avoid layout-per-test. There are several wpt tests Blink has to mark "slow" (gives more than 3 secs budget for the test) or skip because too slow.

**Comment by r12a** [comment]
@mrego I noted your comments on the code, and will bear them in mind for future tests, but i don't have the time to rewrite these tests for things that don't affect the test results or significantly affect the performance, so i'd like to leave the code as is.

File: css/css-text/i18n/ja/css-text-line-break-ja-in-loose.html
Comment (r12a) [reply css/css-text/i18n/ja/css-text-line-break-ja-in-loose.html]: I agree that probably not worth testing Manichaean in this case. @xfq could you remove that line? (except for the `, of course).

**Review by kojiishi — APPROVED** [review]
Given this is a big improvement and a big PR, and multiple people are waiting for it, I'm approving, hoping that all of us can help improving this further together. For my comments, I can submit a follow up PR if that fails in Blink bots.

If @gsnedders and @mrego are happy with this strategy, please leave comments.

**Comment by gsnedders** [comment]
I defer to @mrego as to whether we should block landing the prior raised issues; if we don't, we should ensure an issue is filed to track them.

**Review by frivoal — COMMENTED** [review]
The comment at the bottom of the test files is incorrect. That does not affect the correctness, but it is confusing and misleading. (see my comment on css-text-line-break-ja-cj-loose.html for details)

Other than that, all tests seem correct to me,  but some are testing conditions more narrow than what the spec requires, and some are missing outright. This does not have to block this PR, since they could be added later, but adding them now seems better:

* the -cj- tests are defined for lang=ja, but the spec does not make this behavior depend on language. The tests proposed are correct, but I would suggest adding variants of the same tests:
    * with lang set to zh
    * with lang set to an arbitrary language other than ja or zh
    * with lang unset.

* the -in- and -iteration- tests are defined for lang=ja and lang=zh, but the spec does not make this behavior depend on language. The tests proposed are correct, but I would suggest adding variants of the same tests:
    * with lang set to an arbitrary language other than ja or zh
    * with lang unset.


* the -cpm- and -po- tests check that the line break is allowed for ((ja or zh) and loose), and forbidden for ((ja or zh) and (normal or strict)). That is correct, but the spec also calls for the break to be forbidden if the language is neither ja nor zh, regardles of loose/normal/strict. I would suggest adding a variant of the same tests (changing the expectation in the case of loose):
    * with lang set to an arbitrary language other than ja or zh
    * with lang unset.

* same remark for the -hyphens tests (except that allowed is loose+normal, instead of just loose)

* There is no -pr- test, but I cannot see any reason to omit these.

**Comment by mrego** [comment]
If the rest are fine, I'm ok too. I was just pointing out several issues on the tests that I found,
probably for future tests it'd be nice to get a first single test for reviewing this kind of stuff
and once everyone is happy then replicate the same for the rest of cases.

Still I believe as @gsnedders that we should file an issue about improving these tests
in the future at some point, as having so many duplicated code is going to be a pain for maintenance.

File: css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html
Comment (mrego) [reply css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html]: So this is a personal choice, but the tests are shared with all the browser vendors and the community.
I know there's no a coding style guide for WPT, but I haven't seen this kind of indentation
in the past, and it's not used in the rest of the css-text test suite.
So for me at least is quite confusing. :smile:

File: css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html
Comment (mrego) [reply css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html]: I understand we have a loop for setting up the test and a 2nd loop for actually checking things,
that way we avoid multiple layouts.

But I was talking about the last 2 loops, I don't see the benefit of having those split in 2.

Anyway it seems this kind of issues should be addressed in the future.

File: css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html
Comment (r12a) [reply css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html]: @xfq could you replace "The test creates a box width that causes the 8th character to wrap." with the text that Florian suggested? (perhaps with a line-break before the remainder of the comment, for readability)

File: css/css-text/i18n/ja/css-text-line-break-ja-in-loose.html
Comment (xfq) [reply css/css-text/i18n/ja/css-text-line-break-ja-in-loose.html]: Done.

File: css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html
Comment (xfq) [reply css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html]: Done.

**Comment by kojiishi** [comment]
I think I captured all feedback in #11161, but please add/modify if I missed or interpreted incorrectly. And also contributions to fix is even more appreciated!

**Comment by r12a** [comment]
Thanks, Koji, all.

**Comment by kojiishi** [comment]
Thank you too, Richard!!

File: css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html
Comment (kojiishi) [reply css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html]: I think the current string format is easier to maintain when upgrading Unicode versions, since most Unicode data is provided in text files. I agree your suggestion is easier for JavaScript to read, but then someone must translate the source text data to JSON. Isn't it better for computer to handle such machinery translations?

Comment (mrego) [reply css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html]: Of course, if that's copy&pasted from the Unicode spec, then it makes all the sense to have it like a string.
I didn't know it, that's why I was suggesting something different. Thanks for the explanation.

Comment (gsnedders) [reply css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html]: Can we have some comment saying exactly where the charlist comes from (so others can update it, etc.)?

Comment (frivoal) [reply css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html]: The actual lists comes from Unicode, but what part of unicode to copy and paste is described in the spec.

> breaks before Japanese small kana or the Katakana-Hiragana prolonged sound mark: i.e. characters with the Unicode Line Break property CJ. (See LineBreak.txt in [UNICODE].) 

we could copy that sentence, link included, into the test, but going through the spec, to which the text does link, may be better to make sure you don't miss something if the spec is changed somehow.

Comment (r12a) [reply css/css-text/i18n/ja/css-text-line-break-ja-cj-loose.html]: I used this for the main data source: https://www.unicode.org/Public/UNIDATA/extracted/DerivedLineBreak.txt, expanding the ranges and reformatting to get the list in the tests. For more information about expected line break behavior and line break classes, see [Unicode Standard Annex #14 Line Breaking Properties](http://www.unicode.org/reports/tr14/).

**Pull request merged**
